# DATA WRANGLING AND TIDYING
## EDA: Diagnosing Diabetes

In this project, you’ll imagine you are a data scientist interested in exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

You will use your EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database)
    is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration a 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure (mm Hg)
- `SkinThickness`: Triceps skinfold thickness (mm)
- `Insulin`: 2-Hour serum insulin (muU/ml)
- `BMI`: Body mass index (kg/m^2)
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

Let’s get started!

### Initial Inspection

1. First, familiarize yourself with the dataset
    [here](https://www.kaggle.com/uciml/pima-indians-diabetes-database).

    Look at each of the nine columns in the documentation.

    What do you expect each data type to be?

    *Answer*: It looks like all `dtype`s should be `int`s or `float`s.

2. Next, let’s load in the diabetes data to start exploring.

    Load the data in a variable called `diabetes_data` and print the first few rows.

    **Note**: The data is stored in a file called `diabetes.csv`.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_rows", None)

diabetes_data = pd.read_csv("diabetes.csv")

print(diabetes_data.info())
diabetes_data.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

3. How many columns (features) does the data contain?

   *Answer*: 9

4. How many rows (observations) does the data contain?

    *Answer*: 768

5. Let’s inspect diabetes_data further.

    Do any of the columns in the data contain null (missing) values?

In [2]:
diabetes_data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

*Answer*: Looks like no...

6. If you answered no to the question above, not so fast!

    While it’s technically true that none of the columns contain null values, that doesn’t necessarily mean that the data isn’t missing any values.

    When exploring data, you should always question your assumptions and try to dig deeper.

    To investigate further, calculate summary statistics on `diabates_data` using the `.describe()` method.

In [3]:
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

7. Looking at the summary statistics, do you notice anything odd about the following columns?
    - `Glucose`
    - `BloodPressure`
    - `SkinThickness`
    - `Insulin`
    - `BMI`

*Answer*: The min value is 0 for them, even though that value should be impossible.

8. Do you spot any other outliers in the data?

*Answer*: Max insulin is 846, max pregnancies is 17

9. Let’s see if we can get a more accurate view of the missing values in the data.

    Use the following code to replace the instances of `0` with `NaN` in the five columns mentioned:
    
    ```python
    diabetes_data[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']] = diabetes_data[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']].replace(0,np.NaN)
    ```

In [4]:
columns = ['Glucose','BloodPressure','SkinThickness','Insulin','BMI']
diabetes_data[columns] = diabetes_data[columns].replace(0, np.NaN)

10. Next, check for missing (null) values in all of the columns just like you did in Step 5.

    Now how many missing values are there?

In [5]:
diabetes_data.isnull().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

11. Let’s take a closer look at these rows to get a better idea of why some data might be missing.

    Print out all of the rows that contain missing (null) values.

In [6]:
diabetes_data[diabetes_data.isnull().any(axis=1)]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
9              8    125.0           96.0            NaN      NaN   NaN   
10             4    110.0           92.0            NaN      NaN  37.6   
11            10    168.0           74.0            NaN      NaN  38.0   
12            10    139.0           80.0            NaN      NaN  27.1   
15             7    100.0            NaN            NaN      NaN  30.0   
17             7    107.0           74.0            NaN      NaN  29.6   
21             8     99.0           84.0            NaN      NaN  35.4   
22             7    196.0           90.0            NaN      NaN  39.8   
23             9    119.0           80.0           35.0      NaN  29.0   
26             7    147.0           76.0            NaN      NaN  39.4   
29             5    117.0           92.0            NaN      NaN  34.1   
30             5    109.0           75.0           26.0      NaN  36.0   
33             6     92.0           92.0            NaN      NaN  19.9   
34            10    122.0           78.0           31.0      NaN  27.6   
36            11    138.0           76.0            NaN      NaN  33.2   
37             9    102.0           76.0           37.0      NaN  32.9   
38             2     90.0           68.0           42.0      NaN  38.2   
41             7    133.0           84.0            NaN      NaN  40.2   
42             7    106.0           92.0           18.0      NaN  22.7   
44             7    159.0           64.0            NaN      NaN  27.4   
45             0    180.0           66.0           39.0      NaN  42.0   
46             1    146.0           56.0            NaN      NaN  29.7   
47             2     71.0           70.0           27.0      NaN  28.0   
48             7    103.0           66.0           32.0      NaN  39.1   
49             7    105.0            NaN            NaN      NaN   NaN   
55             1     73.0           50.0           10.0      NaN  23.0   
58             0    146.0           82.0            NaN      NaN  40.5   
60             2     84.0            NaN            NaN      NaN   NaN   
61             8    133.0           72.0            NaN      NaN  32.9   
62             5     44.0           62.0            NaN      NaN  25.0   
64             7    114.0           66.0            NaN      NaN  32.8   
65             5     99.0           74.0           27.0      NaN  29.0   
66             0    109.0           88.0           30.0      NaN  32.5   
67             2    109.0           92.0            NaN      NaN  42.7   
72            13    126.0           90.0            NaN      NaN  43.4   
74             1     79.0           75.0           30.0      NaN  32.0   
75             1      NaN           48.0           20.0      NaN  24.7   
76             7     62.0           78.0            NaN      NaN  32.6   
77             5     95.0           72.0           33.0      NaN  37.7   
78             0    131.0            NaN            NaN      NaN  43.2   
79             2    112.0           66.0           22.0      NaN  25.0   
80             3    113.0           44.0           13.0      NaN  22.4   
81             2     74.0            NaN            NaN      NaN   NaN   
83             0    101.0           65.0           28.0      NaN  24.6   
84             5    137.0          108.0            NaN      NaN  48.8   
86            13    106.0           72.0           54.0      NaN  36.6   
89             1    107.0           68.0           19.0      NaN  26.5   
90             1     80.0           55.0            NaN      NaN  19.1   
93       

12. Go through the rows with missing data. Do you notice any patterns or overlaps between the missing data?

*Answer*: Most rows with missing data are also missing `Insulin` data.

In [7]:
print(diabetes_data.dtypes)

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


13. Next, take a closer look at the data types of each column in diabetes_data.

    Does the result match what you would expect?

*Answer*: Yes, all values are `float`s or `int`s.

**Note**: In Codecademy, the `Outcome` column shows up as `object` `dtype`. This is because some of the `0`s are actually the letter `O`. The data has likely been updated to fix this issue, so we'll skip ahead.

### Next Steps

16. Congratulations! In this project, you saw how EDA can help with the initial data inspection and cleaning process. This is an important step as it helps to keep your datasets clean and reliable.

    Here are some ways you might extend this project if you’d like:

    * Use .value_counts() to more fully explore the values in each column.
    * Investigate other outliers in the data that may be easily overlooked.
    * Instead of changing the 0 values in the five columns to NaN, try replacing the values with the median or mean of each column.

In [8]:
diabetes_data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
3              1     89.0           66.0           23.0     94.0  28.1   
4              0    137.0           40.0           35.0    168.0  43.1   
5              5    116.0           74.0            NaN      NaN  25.6   
6              3     78.0           50.0           32.0     88.0  31.0   
7             10    115.0            NaN            NaN      NaN  35.3   
8              2    197.0           70.0           45.0    543.0  30.5   
9              8    125.0           96.0            NaN      NaN   NaN   
10             4    110.0           92.0            NaN      NaN  37.6   
11            10    168.0           74.0            NaN      NaN  38.0   
12            10    139.0           80.0            NaN      NaN  27.1   
13             1    189.0           60.0           23.0    846.0  30.1   
14             5    166.0           72.0           19.0    175.0  25.8   
15             7    100.0            NaN            NaN      NaN  30.0   
16             0    118.0           84.0           47.0    230.0  45.8   
17             7    107.0           74.0            NaN      NaN  29.6   
18             1    103.0           30.0           38.0     83.0  43.3   
19             1    115.0           70.0           30.0     96.0  34.6   
20             3    126.0           88.0           41.0    235.0  39.3   
21             8     99.0           84.0            NaN      NaN  35.4   
22             7    196.0           90.0            NaN      NaN  39.8   
23             9    119.0           80.0           35.0      NaN  29.0   
24            11    143.0           94.0           33.0    146.0  36.6   
25            10    125.0           70.0           26.0    115.0  31.1   
26             7    147.0           76.0            NaN      NaN  39.4   
27             1     97.0           66.0           15.0    140.0  23.2   
28            13    145.0           82.0           19.0    110.0  22.2   
29             5    117.0           92.0            NaN      NaN  34.1   
30             5    109.0           75.0           26.0      NaN  36.0   
31             3    158.0           76.0           36.0    245.0  31.6   
32             3     88.0           58.0           11.0     54.0  24.8   
33             6     92.0           92.0            NaN      NaN  19.9   
34            10    122.0           78.0           31.0      NaN  27.6   
35             4    103.0           60.0           33.0    192.0  24.0   
36            11    138.0           76.0            NaN      NaN  33.2   
37             9    102.0           76.0           37.0      NaN  32.9   
38             2     90.0           68.0           42.0      NaN  38.2   
39             4    111.0           72.0           47.0    207.0  37.1   
40             3    180.0           64.0           25.0     70.0  34.0   
41             7    133.0           84.0            NaN      NaN  40.2   
42             7    106.0           92.0           18.0      NaN  22.7   
43             9    171.0          110.0           24.0    240.0  45.4   
44             7    159.0           64.0            NaN      NaN  27.4   
45             0    180.0           66.0           39.0      NaN  42.0   
46             1    146.0           56.0            NaN      NaN  29.7   
47             2     71.0           70.0           27.0      NaN  28.0   
48             7    103.0           66.0           32.0      NaN  39.1   
49             7    105.0            NaN            NaN      NaN   NaN   
50             1    103.0           80.0           11.0     82.0  19.4   
51             1    101.0           50.0           15.0     36.0  24.2   
52             5     88.0           66.0           21.0     23.0  24.4   
53       